### ISSUE TITLE
Substring searches give non-deterministic output

### ISSUE THREAD

**Issue by reuster986** [mhmerrill/arkouda#301]
String methods `.contains`, `.startswith`, and `.endswith` are giving non-deterministic answers on large data in multi-locale settings. In `tests/string_test.py` on single-locale, their answers pass the tests. There might be a race condition somewhere.

**Comment by reuster986** [comment]
Actually, `.startswith` is giving deterministic output; it's only the other two that aren't. This might be related to another issue, that the performance of `.contains` and `.endswith` are much slower than `.startswith`. The two functions that are slow and non-deterministic use domain slicing here, while the fast, deterministic one does not:
https://github.com/mhmerrill/arkouda/blob/59da8f05f8dbf71382083964bc1b59ddceedc1ac/src/SegmentedArray.chpl#L374-L390

@bradcray and @ronawho I can easily see why domain slicing would slow things down, but is there a way that it could be causing a race condition too?

**Comment by bradcray** [comment]
There's nothing about slices that would inherently cause a race condition, though races could be written using slices (if, for example, a slice that was being read aliased another slice that was simultaneously being written).  With a quick glance at your code, it doesn't look like that should be the case, though right?  (I'm focusing in particular on the `endsWith` case where I believe `hits` does not alias with `truth` or `oa`?

What's the nature of the nondeterministic output... sometimes the right answer and sometimes garbage?  or something else?  And can I easily reproduce this for myself somehow?  (e.g., is running string_test.py on multiple locales sufficient?)

**Comment by reuster986** [comment]
Update: I fixed the non-deterministic output by replacing the domain slicing with calls to `oD.interior`, e.g. in `endsWith`:

```chapel
hits[oD.interior(-(oD.size-1))] = truth[oa[oD.interior(oD.size-1)] - substr.numBytes - 1];
```

The nondeterministic behavior was that an arkouda command of the form `strings.endswith(substring).sum()`, when executed multiple times in a row, would give slightly different values each time. Not wildly different, but like 0.001% off.

I think you're right that there is no aliasing going on, so I don't know why domain slicing would cause this. Unfortunately, because arkouda does not yet have a function for generating random arrays of strings, it's difficult to reproduce this behavior without reading in a lot of actual data.

But in the course of tracking down this issue, I found another (presumably related) source of the same kind of non-deterministic behavior: performing a gather on boolean arrays with a non-unique index vector *sometimes* gives non-deterministic results, seemingly only after allocating a lot of memory (say, 5 GB per locale). The following arkouda code illustrates:

```python
Nvals = 10**6
Ninds = 10*Nvals
values = ak.randint(0, 2, Nvals, dtype=VALTYPE)
inds = ak.randint(0, Nvals, Ninds, dtype=ak.int64)
pops = [values[inds].sum() for trial in range(10)]
Nanswers = len(set(pops)) # should be 1
```

When `VALTYPE` is `ak.int64`, this code reliably gives one answer, but when `VALTYPE` is `ak.bool`, it sometimes gives more than one answer, under certain conditions that I can't quite pin down. The relevant chapel code for gather is here:

https://github.com/mhmerrill/arkouda/blob/59da8f05f8dbf71382083964bc1b59ddceedc1ac/src/IndexingMsg.chpl#L114-L136

Could there be something going on with aggregation of bools, similar to #236 ?

**Comment by bradcray** [comment]
@reuster986: This won't be reassuring, but I can't think of any reason that using the domain slicing calls with interior would help...  those routines are essentially just helpers that create domains the same way you did.  Is the behavior really so night and day that you have great confidence it's gone due to that change?

Given both the original mystery and the new int vs. bool one, I find myself wondering whether the race could be in .sum() rather than in the slicing.  Dumb question, but: Is that an Arkouda method or a Chapel library one?  With a quick look, I'm not finding it in either source base, so am obviously missing something.

**Comment by bradcray** [comment]
(oh, and: If the source of the int vs. bool bug is in the aggregation, I'll probably need @ronawho's help to look into it...)

**Comment by ronawho** [comment]
@reuster986 is there a full reproducer and do you know what platform(s) you're seeing issues on? (XC, IB cluster, locally, etc) The output of ` printchplenv --anonymize` may be helpful too.

I tried the following on 16 nodes of an XC using arkouda master built with a patched version of 1.20 (1.20 with https://github.com/chapel-lang/chapel/pull/14652 applied):


```python
#!/usr/bin/env python3

import sys

import arkouda as ak

ak.verbose = False
ak.connect(server=sys.argv[1], port=sys.argv[2])

VALTYPE = ak.bool
Nvals = 10**7
Ninds = 10*Nvals
values = ak.randint(0, 2, Nvals, dtype=VALTYPE)
inds = ak.randint(0, Nvals, Ninds, dtype=ak.int64)
pops = [values[inds].sum() for trial in range(10)]
Nanswers = len(set(pops))
print(Nanswers)

ak.disconnect()
```

I ran a couple dozen times and only ever saw Nanswers=1. If you're on an XC and do not have the fix for https://github.com/mhmerrill/arkouda/issues/236 applied, I would not be surprised if you're seeing random wrong answers. The root problem for that bug is that with 1.20 if you have a GET whose size or address isn't 4-byte aligned we end up performing way too large of a mempcy that can overwrite arbitrary user data.

**Comment by reuster986** [comment]
Sorry for the delay, guys. After much testing, I still have not found a way to reliably reproduce this behavior. I have a script that was consistently giving non-deterministic results from bool gather on a particular 16 nodes of an XC, but then I ran it again an hour later on a different 16-node allocation and everything was correct. On the first allocation where the non-determinism was happening, I established that the bug would only trigger if I first read in a large (~100 GB) array of strings and computed `sipHash128` of each string. Then, if I performed a gather on a bool array as above, with `Nvals=10**7` and `Ninds=10**8`, the answer was different across several trials. 

I used a modified test that omitted the `.sum()` and just directly compared the results of two identical gathers, to rule out the possiblity that `.sum()` was the culprit.

```python
def test(inputSize, outputSize, dt, trials):
    bag = ak.randint(0, 2, inputSize, dtype=dt)
    inds = ak.randint(0, inputSize, outputSize, dtype=ak.int64)
    res = []
    for trial in range(trials):
        t1 = bag[inds]
        t2 = bag[inds]
        res.append(((1*t1) == (1*t2)).all())
    return res
```

(The `1*t1` is necessary to cast t1 to an int64 array, because arkouda does not yet implement `==` for bool arrays).

I confirmed that I am using a patched version of Chapel 1.20 with the bool-concatenate bug fixed.

I still have not been able to reproduce this behavior on a non-XC system, although the only other platform I've tried is an SMP.

I know this doesn't really give you guys much to go on, but do you see any clues in here? Could it be a hardware problem with particular nodes in our system? Or maybe there is still an undiscovered problem with sub-word GETs, and the code in `SipHash.chpl` is triggering it when it localizes the bytes to be hashed?

Tomorrow I will add functionality to generate random arrays of strings from the arkouda client, so that we can test things like this more easily. I will also try to reproduce this bug on an IB cluster or my laptop in multi-locale mode.

**Comment by ronawho** [comment]
This sounds like it could be a memory corruption issue, similar to the other bug where we might be overwriting memory we shouldn't be. If you can send me a reproducer that sometimes fails I can try a few things/tools to make those types of failures more deterministic.

**Comment by reuster986** [comment]
I created a reproducer that sometimes fails on [this branch](https://github.com/mhmerrill/arkouda/tree/bool-gather-reproducer) at `tests/bool_gather.py`. That branch pulls in my latest two PRs to be able to generate arrays of random strings. The default settings do 10 trials, where each trial is a gather of 10**7 bools (from an array of size 10**6). These parameters can be set on the command line. I chose the default gather size so that it would run on my laptop, but in order to trigger the bug on an XC with 256 GB/locale, I had to use at least `--input-size=10**7` and `--output-size=10**8`.

**Comment by bradcray** [comment]
Bill, thanks for putting in the effort to create a reproducer.  You mentioned that you hadn't seen this fail on a non-XC system, is that still true?  Have you had the opportunity to try it on an XC system, but using a communication layer other than `ugni` by any chance?  (e.g., gasnet/mpi or gasnet/aries?)

**Comment by ronawho** [comment]
Thanks, I'll see if I can reproduce the failure.

For future reference here's how I'm building/running:
```
git clone https://github.com/mhmerrill/arkouda.git --depth=1 --branch=bool-gather-reproducer arkouda-bool-gather-reproduce
cd arkouda-bool-gather-reproducer/
make -j install-deps
make
pip3 install -e . --user

./util/test/runClient tests/bool_gather.py --input-size=$((10**7)) --output-size=$((10**8)) -nl 16

> connected to arkouda server tcp://*:5555
> Got expected result
```

I'll start this in a loop and run with some tools that might help identify invalid memory accesses.

Would you mind sending along the output of `printchplenv --anonymize`,  `module list -t`, and `chpl --version` (can be off-list if that's easier)

**Comment by reuster986** [comment]
@bradcray I've tested on my laptop and got the normal/correct behavior, but it was with `CHPL_COMM=none`. I have a couple things to do before switching over to `gasnet` on my laptop and testing with that. Meanwhile, I'm building arkouda on an IB cluster to test there.

However, as I'm writing this, I'm realizing that both my laptop and the IB cluster have unpatched Chapel 1.20. Will that matter, or does the patch only affect ugni?

I don't think our XC has Chapel built with any comm layer other than ugni. If that test would be valuable enough to warrant a separate build, I would probably have to ask @BryantLam to build Chapel with the alternate comm layers, since he's way more familiar than me with building Chapel on that system.

@ronawho unfortunately, I'm not at my work site today, so I can't access the XC until tomorrow to get you the environment info. But I can tell you that I've seen the bug trigger on both 16 and 45 locales (haven't tried anything else).

**Comment by bradcray** [comment]
Hi @reuster986 —

Now that you've provided a reproducer, it's probably easier for us to test non-ugni configurations than for you to worry about it.  I was just curious whether you had any additional data there.

The previous patch for bool transfers was for ugni only, so the fact that your laptop and IB cluster are unpatched should be fine.

**Comment by reuster986** [comment]
@bradcray cool, thanks!

**Comment by reuster986** [comment]
I ran the reproducer several times on 4 simulated locales on my laptop and 4 locales of an IB cluster and have gotten only correct results so far.

**Comment by ronawho** [comment]
I ran `./util/test/runClient tests/bool_gather.py --input-size=$((10**7)) --output-size=$((10**8)) -nl 16` 10K times and have only seen correct results so far. I also did a build with AddressSanitizer, and did not detect any memory errors. 

It'd be helpful if you could:
 - Compile [test/runtime/configMatters/comm/misalignedComm.chpl](https://github.com/chapel-lang/chapel/blob/master/test/runtime/configMatters/comm/misalignedComm.chpl) and verify you get the output `false` to ensure the ugni bug-fix has been properly applied
 - Send the output of  `printchplenv --anonymize`, `module list -t`, and `./arkouda_server --about`
 - Send machine specs:
   - number of nodes used (think you said 16 and 45)
   - number of cores per node and cpu type (could do `srun cat /proc/cpuinfo`)
   - CLE version (`cat /etc/opt/cray/release/CLEinfo 2>/dev/null || cat /etc/opt/cray/release/cle-release`)

Feel free to send that info off-list (on list is fine with me too, but I don't know how sensitive machine specs are)

> Could it be a hardware problem with particular nodes in our system

We are very far from the experts on this aspect :) If there's a node in a bad state, I usually try to bisect and start with all nodes and just keep halving in different directions to find the responsible one. I don't know how practical that is for you though.

**Comment by reuster986** [comment]
I collected this information from a terminal session that successfully reproduced the bug with `tests/bool_gather.py` and an `arkouda_server` on 45 locales.
* Running `misalignedComm`, I get `false` with both `-nl 16` and `-nl 45`.
* I'll send you an email with output of `printchplenv --anonymize` and 1module list -t`.
* I've tested on 16 and 45 nodes. Sending you machine specs in email.
  
With slurm, how do you select particular nodes to run on?

**Comment by ronawho** [comment]
> With slurm, how do you select particular nodes to run on?

There's a couple ways, but the easiest is probably to throw `--nodelist` to the launcher wrapper. Something like:

```
./arkouda_server -nl 45 --nodelist=nid00[00-44]
./arkouda_server -nl 23 --nodelist=nid00[00-22]
./arkouda_server -nl 12 --nodelist=nid00[00-11]
```

This maps to `srun --nodelist` so see --nodelist in https://slurm.schedmd.com/srun.html for more info.

There's also `--exclude` if you think you've found bad node(s), and want to avoid just them, but run on everything else.

**Comment by ronawho** [comment]
Oh, and if it's easier to do with env vars you can instead use `CHPL_LAUNCHER_NODELIST` / `CHPL_LAUNCHER_EXCLUDE`

**Comment by ronawho** [comment]
Hmm, I was able to reproduce this once at 256 nodes with a larger problem size (`--input-size=$((10**8)) --output-size=$((10**9))`), but I cannot reliably produce the failure. Still that should be enough for me to go on (and likely rule out hardware issues.)

**Comment by reuster986** [comment]
Well, I'm relieved that I'm not crazy! Thank you for digging into this.

**Comment by ronawho** [comment]
Some misc notes (that are mostly for myself):

For a GET, the Aries NIC requires that the remote address and the size are 4-byte aligned. Most of our benchmarks tend to use 8-byte types, so our GETs tend to be naturally 4-byte aligned. When a GET has a misaligned size or address, we over-GET into bounce buffers. (We do a 4-byte-aligned GET that is larger than originally requested into some different memory and then memcpy the correct bytes into the original buffer.)

Something like:

```c

void* bounce_buffer_address = get_bounce_buffer(aligned_size);
do_nic_get(bounce_buffer_address, remote_node, 4_byte_align_down(remote_source_address), aligned_size);
memcpy(orig_target_address, bounce_buffer_address+misaligned_amount, orig_size);
free_bounce_buffer(bounce_buffer_address)
```

This is not a code-path that gets stressed regularly, but you are using it pretty heavily when doing single-byte (bool) GETs. Currently the aggregators under ugni still map to unorderedCopy for GETs, which ends up mapping to code like above (unorderedCopy for misaligned GETs is implemented with ordered GETs)

---

We're going to look at a couple things:
 - Find the cause of this bug (current best guess is a subtle race acquiring a bounce buffer)
 - Use aggregation for GETs under ugni (in the current implementation too many tasks will be created under ugni in 1.20 at scale, which can cause OOMs. We worked around that for PUTs in https://github.com/mhmerrill/arkouda/pull/209, but still have to do that for GETs.) Assuming we're right about the source of the bug, this will work-around the bug in 1.20. It will also improve performance of GETs, especially these smaller ones that end up using ordered comm.
 - Optimize misaligned GETs under ugni. If we switch to aggregation this won't be as important for most code-paths, but if you're doing bulk transfers of array-slices it could still matter. e..g if you're doing something like `A[1..999999] = B[1..999999]` for boolean arrays, this will optimize that.

**Comment by ronawho** [comment]
I was able to reproduce pretty consistently by decreasing the size of the bounce-buffer pool, and I think I've found the source of the bug (big thanks to @gbtitus for the recommendation to look at the bounce buffer code). Here's a patch that should resolve the problem:


```diff
diff --git a/runtime/src/comm/ugni/comm-ugni.c b/runtime/src/comm/ugni/comm-ugni.c
index 83e619a084..794b27bf6d 100644
--- a/runtime/src/comm/ugni/comm-ugni.c
+++ b/runtime/src/comm/ugni/comm-ugni.c
@@ -6120,7 +6120,7 @@ void do_remote_get(void* tgt_addr, c_nodeid_t locale, void* src_addr,
     //
     // The transfer will fit in a single trampoline buffer.
     //
-    tgt_addr_xmit = get_buf_alloc(size);
+    tgt_addr_xmit = get_buf_alloc(xmit_size);

     do_nic_get(tgt_addr_xmit, locale, remote_mr,
                src_addr_xmit, xmit_size, gnr_mreg);
```

I'll see if I can come up with a simpler and more consistent standalone Chapel reproducer, and will plan to run `tests/bool_gather.py` in a loop over the weekend with this patch applied.

**Comment by ronawho** [comment]
I'm pretty confident that bug is the source of the problem. Basically what happens is we're grabbing too small of a buffer and overflowing it. If we're trying to do a 7-byte GET on an address that has 7-byte alignment, we'll end up doing a 16-byte GET, but we were only asking for space to store 8-bytes. We'd overflow that 8-byte buffer, which corrupts data structures around it, which are later needed by the 1-byte GETs.

For standalone 1-byte GETs, the max size needed for any alignment is 8-bytes, which is why the bool gather itself wasn't enough to trigger the bug. The pretest is needed to trigger GETs with things like 7-byte sized 7-byte aligned GETs to first overflow and corrupt some of the data structures.

I opened https://github.com/chapel-lang/chapel/pull/15111 to fix this bug upstream, and that includes a standalone reproducer. 

The patch in https://github.com/mhmerrill/arkouda/issues/301#issuecomment-595713142 should apply to 1.20. If you can apply that patch to your builds, I think that will be the cleanest solution. If not we can probably come up with a workaround in the module code to avoid initiating misaligned GETs (the only place that I know of that can cause this currently is the manual GET in SipHash)

**Comment by reuster986** [comment]
@ronawho thank you for identifying this and for the very helpful explanation of why the pretest was necessary. I had hypothesized that something like that was happening, but I felt a little crazy writing that reproducer. Anyway, I will talk with @BryantLam about patching the Chapel build. Thanks again!

**Comment by ronawho** [comment]
Yeah racy runtime bugs (especially in the comm layer) are often able to convince you that you're insane -- https://xkcd.com/2259/

**Comment by reuster986** [comment]
Verified that this fixes the bug in our use cases. Thanks @ronawho !
